File: src/typescript/src/transport/transaction/message/incoming-message.ts
import { IPolyBus } from '../../../i-poly-bus';
import { Message } from './message';
import { MessageInfo } from './message-info';

/**
 * Represents an incoming message in the transport layer.
 * Contains the message body, deserialized message object, and message type information.
 */
export class IncomingMessage extends Message {
  private _messageInfo: MessageInfo;
  private _messageType: any;
  private _body: string;
  private _message: any;

  /**
   * Creates a new IncomingMessage instance.
   * @param bus The bus instance associated with the message.
   * @param body The message body contents.
   * @param messageInfo The message info describing metadata about the message.
   */
  constructor(bus: IPolyBus, body: string, messageInfo: MessageInfo) {
    super(bus);

    if (!body) {
      throw new Error('Body parameter cannot be null or undefined');
    }

    if (!messageInfo) {
      throw new Error('MessageInfo parameter cannot be null or undefined');
    }

    this._body = body;
    this._message = body;
    this._messageInfo = messageInfo;
    this._messageType = bus.messages.getTypeByMessageInfo(messageInfo);
  }

  /**
   * The message info describing metadata about the message.
   */
  public get messageInfo(): MessageInfo {
    return this._messageInfo;
  }

  public set messageInfo(value: MessageInfo) {
    this._messageInfo = value;
  }

  /**
   * The default is string, but can be changed based on deserialization.
   */
  public get messageType(): any {
    return this._messageType;
  }

  public set messageType(value: any) {
    this._messageType = value;
  }

  /**
   * The message body contents.
   */
  public get body(): string {
    return this._body;
  }

  public set body(value: string) {
    this._body = value;
  }

  /**
   * The deserialized message object, otherwise the same value as Body.
   */
  public get message(): any {
    return this._message;
  }

  public set message(value: any) {
    this._message = value;
  }
}

File: src/python/src/transport/transaction/message/incoming_message.py
from typing import Any, Optional, Type
from .message import Message
from .message_info import MessageInfo
from ....i_poly_bus import IPolyBus


class IncomingMessage(Message):
    """
    Represents an incoming message from the transport.
    """

    def __init__(self, bus: "IPolyBus", body: str, message_info: MessageInfo):
        super().__init__(bus)
        if body is None:
            raise ValueError("body cannot be None")
        if message_info is None:
            raise ValueError("message_info cannot be None")
        
        self._message_info = message_info
        self._message_type = bus.messages.get_type_by_message_info(message_info)
        self._body = body
        self._message = body

    @property
    def message_info(self) -> MessageInfo:
        """
        The message info describing metadata about the message.
        """
        return self._message_info

    @message_info.setter
    def message_info(self, value: MessageInfo) -> None:
        """
        Set the message info.
        """
        self._message_info = value

    @property
    def message_type(self) -> Type:
        """
        The default is string, but can be changed based on deserialization.
        """
        return self._message_type

    @message_type.setter
    def message_type(self, value: Type) -> None:
        """
        Set the message type.
        """
        self._message_type = value

    @property
    def body(self) -> str:
        """
        The message body contents.
        """
        return self._body

    @body.setter
    def body(self, value: str) -> None:
        """
        Set the message body contents.
        """
        self._body = value

    @property
    def message(self) -> Any:
        """
        The deserialized message object, otherwise the same value as Body.
        """
        return self._message

    @message.setter
    def message(self, value: Any) -> None:
        """
        Set the deserialized message object.
        """
        self._message = value
